cmd/fs_handlers.go

```diff
@@ -96,13 +93,25 @@ func walk(root, repoRoot string, depth int) ([]stagePair, error) {
 				if err != nil {
 					return err
 				}
-				toBeStaged = append(toBeStaged, extra...)
+				for k, v := range extra {
+					list, ok := toBeStaged[k]
+					if !ok {
+						list = []string{}
+					}
+					list = append(list, v...)
+					toBeStaged[k] = list
+				}
 				return nil
 			}
 		}
 
 		if info.Mode().IsRegular() {
-			toBeStaged = append(toBeStaged, stagePair{childPath, repoPath})
+			list, ok := toBeStaged[childPath]
```

Comment:
I'm confused. You're defining equalness based on the path. This will work only in case of absolute symbolic links. Sadly, the more common form I saw are relative symbolic links. Those will lead to different paths for the same file (did I mention that I hate the concept?):

```bash
→ touch im-the-same-file-always
→ ln -s . sub-directory
→ cd sub-directory
→ pwd
/tmp/t/sub-directory/sub-directory
→ stat /tmp/t/sub-directory/sub-directory/im-the-same-file-always
 File: /tmp/t/sub-directory/sub-directory/im-the-same-file-always
 Device: 30h/48d	Inode: 61479275    Links: 1
 ...
→ cd sub-directory
→ stat /tmp/t/sub-directory/sub-directory/sub-directory/im-the-same-file-always
 File: /tmp/t/sub-directory/sub-directory/sub-directory/im-the-same-file-always
 Device: 30h/48d	Inode: 61479275    Links: 1
 ...
```

As you can see, the path changes since it's totally valid for the same file to have several paths (-> hardlinks). The only thing that stays constant is the combination of ``device/inode`` number. You should use this as key to uniqueness.
```

Reply:
> I'm confused. You're defining equalness based on the path. This will work only in case of absolute symbolic links.

If we track only symbolic links to files, the path equality is good enough.
`filepath.EvalSymlinks` resolves the link to its absolute path so we are covered.

Tracking links to directories is hard as you correctly pointed out. I am working on it in a new set of patches. I think `EvalSymlinks` still would resolve to the absolute parent, but I need more tests for this case. Current, algorithm reject simple infinite loops due to depth control. But I want to have directory reflinking for the *non abusive* loop free symlinks.  

Hopefully, a user is not in self destruction/harm mode, and does not do loop symlinks unless there is a good case for it. I myself cannot come up with a useful case for looped symlinks.

Reply:
> If we track only symbolic links to files, the path equality is good enough.
>  filepath.EvalSymlinks resolves the link to its absolute path so we are covered.

True. Just tested, and it seems that ``filepath.EvalSymlinks`` is clever enough to always return ``/tmp/t//im-the-same-file-always``. It would be interesting to see how it does that, will check. I still suggest to use dev/inode instead of path due to the added benefit of detecting hardlinks.

> Hopefully, a user is not in self destruction/harm mode, and does not do loop symlinks unless there is a good case for it. I myself cannot come up with a useful case for looped symlinks.

There is probably none. It's "just" a security thing from my perspective.